rust/tw_ss58_address/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+use blake2::{Blake2b512, Digest};
+use std::fmt::Formatter;
+use std::str::FromStr;
+use tw_coin_entry::error::prelude::*;
+use tw_encoding::{base58, hex};
+use tw_keypair::tw::{PrivateKey, PublicKey};
+
+//
+// Most of the materials implemented here is based on the following resources:
+// - https://wiki.polkadot.network/docs/learn-account-advanced#address-format
+// - https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/core/src/crypto.rs
+//
+
+#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub struct NetworkId(u16);
+
+impl NetworkId {
+    fn new_unchecked(value: u16) -> Self {
+        return Self(value);
+    }
+
+    pub fn from_u16(value: u16) -> AddressResult<Self> {
+        match value {
+            0..=0x3fff => Ok(Self::new_unchecked(value)),
+            _ => Err(AddressError::InvalidInput),
+        }
+    }
+
+    pub fn value(&self) -> u16 {
+        self.0
+    }
+}
+
+impl TryFrom<u16> for NetworkId {
+    type Error = AddressError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Self::from_u16(value)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct SS58Address {
+    key: Vec<u8>,
+    network: NetworkId,
+}
+
+impl SS58Address {
+    const CHECKSUM_SIZE: usize = 2;
+    const KEY_SIZE: usize = 32;
+    const SS58_PREFIX: &'static [u8] = b"SS58PRE";
+
+    fn extract_network(bytes: &[u8]) -> AddressResult<(usize, u16)> {
+        if bytes.len() == 0 {
+            return Err(AddressError::MissingPrefix);
+        }
+
+        match bytes[0] {
+            0..=63 => Ok((1, bytes[0] as u16)),
+            64..=127 if bytes.len() >= 2 => {
+                let lower = (bytes[0] << 2) | (bytes[1] >> 6);
+                let upper = bytes[1] & 0b0011_1111;
+                Ok((2, (lower as u16) | ((upper as u16) << 8)))
+            },
+            _ => Err(AddressError::UnexpectedAddressPrefix),
+        }
+    }
+
+    fn encode_network(network_id: NetworkId) -> Vec<u8> {
+        let network = network_id.value();
+        match network {
+            0..=63 => {
+                vec![network as u8]
+            },
+            _ => {
+                let first = ((network & 0b0000_0000_1111_1100) as u8) >> 2;
+                let second =
+                    ((network >> 8) as u8) | ((network & 0b0000_0000_0000_0011) as u8) << 6;
+                vec![first | 0b01000000, second]
+            },
+        }
+    }
+
+    fn compute_expected_checksum(decoded: &[u8]) -> Vec<u8> {
+        // XXX: tried using tw_hash::blake2::blake2_b, but it did not produce the correct result
```

Comment:
Did you try to concatenate `SS58_PREFIX` and `decoded` before calling `tw_hash::blake2::blake2_b`?